pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -102,6 +103,8 @@ public boolean add(MessageImpl<?> msg, SendCallback callback) {
                 this.firstCallback = callback;
                 batchedMessageMetadataAndPayload = allocator.buffer(
                         Math.min(maxBatchSize, ClientCnx.getMaxMessageSize()));
+                extraCapacity = batchedMessageMetadataAndPayload.capacity();
+                producer.client.getMemoryLimitController().forceReserveMemory(extraCapacity);
```

Comment:
IIUC, the main idea of this change is to limit allocated memory not just used.
I am not sure if there are better implementations, but can we just reserve memory after allocation and release in `clean()`? , and skip previous reserve/release when it's batched.

Reply:
> but can we just reserve memory after allocation

The buffer we allocated can be resized, so it's not correct if we only reserver memory after allocation

>  and release in clean()

The buffer will be released until  broker ack received, so it looks that we can not release in `clean()` @Jason918